### PR TITLE
makeClone returns a std::unique_ptr<WaveFunctionComponent>

### PR DESCRIPTION
**Pull request by quantumsteve**
Please review the [developer documentation](https://github.com/QMCPACK/qmcpack/wiki/Development-workflow)
on the wiki of this project that contains help and requirements.

## Proposed changes

Describe what this PR changes and why.  If it closes an issue, link to it here
with [a supported keyword](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

This changes makeClone to return a std::unique_ptr<WaveFunctionComponent>. This fixes several memory leaks found by address sanitizer.

This is part of #3083.

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- Bugfix
- Refactoring (no functional changes, no api changes)

### Does this introduce a breaking change?

- No

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes This PR is up to date with current the current state of 'develop'
- Yes Code added or changed in the PR has been clang-formatted
- n/a This PR adds tests to cover any new code, or to catch a bug that is being fixed
- n/a Documentation has been added (if appropriate)

### CODE REVIEW

**Comment by ye-luo** [comment]
Test this please

**Comment by ye-luo** [comment]
As WFC::makeClone is a virtual function. I would like to take one step back. All the WFC derived class needs to have override protection on first. Then we change return type.

File: src/QMCWaveFunctions/Fermion/DiracDeterminantBase.h [hunk 1]
```diff
@@ -133,10 +133,10 @@ class DiracDeterminantBase : public WaveFunctionComponent
   }
 
   // Stop makeClone
-  WaveFunctionComponentPtr makeClone(ParticleSet& tqp) const override final
+  std::unique_ptr<WaveFunctionComponent> makeClone(ParticleSet& tqp) const override
```
Comment (ye-luo) [inline src/QMCWaveFunctions/Fermion/DiracDeterminantBase.h:136]: This was intentionally made final.

**Review by ye-luo — COMMENTED** [review]
It seems that all the modified classes have override protection on.
@quantumsteve could you double check the number of modified makeClone is consistent with the number of classes derived from WFC?

File: src/QMCWaveFunctions/Fermion/DiracDeterminantBase.h (hunk 1 above)
Comment (ye-luo) [reply src/QMCWaveFunctions/Fermion/DiracDeterminantBase.h:136]: Probably to add this comment "final keyword is intended to disable makeClone being further inherited"

Comment (quantumsteve) [reply src/QMCWaveFunctions/Fermion/DiracDeterminantBase.h:136]: OK, I'll change it back to final.

**Comment by quantumsteve** [comment]
> It seems that all the modified classes have override protection on.
> @quantumsteve could you double check the number of modified makeClone is consistent with the number of classes derived from WFC?

I temporarily marked `WaveFunctionComponent::makeClone` as `final` and everything inheriting from it had the `override` keyword.

**Comment by ye-luo** [comment]
> > It seems that all the modified classes have override protection on.
> > @quantumsteve could you double check the number of modified makeClone is consistent with the number of classes derived from WFC?
> 
> I temporarily marked `WaveFunctionComponent::makeClone` as `final` and everything inheriting from it had the `override` keyword.

Nice trick. It is good to check now but I think it needs needs to be done before an API change was attempt. If some override case has been missing before the API change, it will not be caught by "final" the virtual function after the API change.

**Comment by quantumsteve** [comment]
Do we want to merge #3269 first?

**Comment by ye-luo** [comment]
> Do we want to merge #3269 first?

I was wondering too but I had some doubts in that one. So I'd like to merge this one first.

do you have answer for my earlier question
>  @quantumsteve could you double check the number of modified makeClone is consistent with the number of classes derived from WFC?

**Comment by ye-luo** [comment]
Test this please

**Review by ye-luo — APPROVED** [review]
I grepped makeClone of all the classes inherited WaveFunctionComponent. override is on.

**Comment by williamfgc** [comment]
@ye-luo @quantumsteve thanks for wrapping this up. @quantumsteve is it safe to delete my own branch in my fork?

**Comment by quantumsteve** [comment]
@williamfgc Yes, go ahead and delete your branch.

**Pull request merged**